src/main/java/org/myrobotlab/service/Py4j.java

```diff
@@ -1,148 +1,352 @@
+      this.waitFor = new WaitForProcess(py4j, process);
+      this.waitFor.start();
+
+      log.info("process started {}", process);
+    }
+  }
+
+  /**
+   * A class to wait on a process signal and notify client is disconnected
+   * 
+   * @author GroG
+   *
+   */
+  public class WaitForProcess extends Thread {
+    public Integer exitCode;
+    public Process process;
+    public Py4j py4j;
+
+    public WaitForProcess(Py4j py4j, Process process) {
+      super(String.format("%s-process-signal", py4j.getName()));
+      this.process = process;
+      this.py4j = py4j;
+    }
+
+    public void run() {
+      try {
+        exitCode = process.waitFor();
+      } catch (InterruptedException e) {
+      }
+      warn("process %s terminated with exit code %d", process.toString(), exitCode);
+    }
+  }
 
   public final static Logger log = LoggerFactory.getLogger(Py4j.class);
 
+  private static final long serialVersionUID = 1L;
+
+  /**
+   * py4j clients currently attached to this service
+   */
+  protected Map<String, Py4jClient> clients = new HashMap<>();
+
+  /**
+   * Java server side gateway for the python process to attach default port
+   * 25333
+   */
   private transient GatewayServer gateway = null;
 
-  private transient Invoker handler = null;
-  
-  protected HashMap<String, Script> openedScripts = new HashMap<String, Script>();
+  /**
+   * the all important interface to the Python MessageHandler This defines what
+   * Java can call that will rpc'd into Python's MessageHandler
+   */
+  private transient Executor handler = null;
 
-  protected String activeScript = null;
+  /**
+   * Opened scripts are scripts opened in memory, from there they can be
+   * executed or saved to the file system, or updatd in memory which the js
+   * client does
+   */
+  protected HashMap<String, Script2> openedScripts = new HashMap<String, Script2>();
 
+  /**
+   * client process and connectivity reference
+   */
+  protected Py4jClient pythonProcess = null;
+
+  /**
+   * script root directory - all script filenames will be relative to this
+   */
+  protected String scriptRootDir = null;
 
   public Py4j(String n, String id) {
     super(n, id);
   }
 
   /**
-   * start the gateway service listening on port
+   * Add a new script to Py4j default location will be in
+   * data/Py4j/{serviceName}
+   * 
+   * @param scriptName
+   *          - name of the script
+   * @param code
+   *          - code block
+   * @throws IOException
    */
-  public void start() {
-    if (gateway == null) {
-      gateway = new GatewayServer(this);
-      gateway.addListener(this);
-      gateway.start();
-      info("server started listening on %s:%d", gateway.getAddress(), gateway.getListeningPort());
-      handler = (Invoker) gateway.getPythonServerEntryPoint(new Class[] { Invoker.class });
-    } else {
-      log.info("Py4j gateway server already started");
-    }
-  }
+  public void addScript(String scriptName, String code) throws IOException {

```

Comment:
Can this be pulled up into a common interface for both Py4j and Python? Along with the other script-related methods

Reply:
its duplicated in the next pr .. but ya that would be the evolution, or a more general file operations - but the more generalized one will need to wait for some more dependencies to be available...